sdks/python/apache_beam/io/filebasedio_perf_test.py

```diff
@@ -0,0 +1,185 @@
+from apache_beam.testing.test_pipeline import TestPipeline
+from apache_beam.testing.util import assert_that
+from apache_beam.testing.util import equal_to
+
+WRITE_NAMESPACE = 'write'
+READ_NAMESPACE = 'read'
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class FileBasedIOTestOptions(LoadTestOptions):
+  @classmethod
+  def _add_argparse_args(cls, parser):
+    parser.add_argument(
+        '--test_class', required=True, help='Test class to run.')
+    parser.add_argument(
+        '--filename_prefix',
+        required=True,
+        help='Destination prefix for files generated by the test.')
+    parser.add_argument(
+        '--compression_type',
+        default='auto',
+        help='File compression type for writing and reading test files.')
+    parser.add_argument(
+        '--number_of_shards',
+        type=int,
+        default=0,
+        help='Number of files this test will create during the write phase.')
+    parser.add_argument(
+        '--dataset_size',
+        type=int,
+        help='Size of data saved on the target filesystem (bytes).')
+
+
+@typehints.with_output_types(bytes)
+@typehints.with_input_types(Tuple[bytes, bytes])
+class SyntheticRecordToStrFn(beam.DoFn):
+  """
+  A DoFn that convert key-value bytes from synthetic source to string record.
+
+  Output length = 4(ceil[len(key)/3] + ceil[len(value)/3]) + 1
+  """
+  def process(self, element):
+    import base64
+    yield base64.b64encode(element[0]) + b' ' + base64.b64encode(element[1])
+
+
+class CreateFolderFn(beam.DoFn):
+  """Create folder at pipeline runtime."""
+  def __init__(self, folder):
+    self.folder = folder
+
+  def process(self, element):
+    from apache_beam.io.filesystems import FileSystems  # pylint: disable=reimported
+    filesystem = FileSystems.get_filesystem(self.folder)
+    if filesystem.has_dirs() and not filesystem.exists(self.folder):
+      filesystem.mkdirs(self.folder)
+
+
+class TextIOPerfTest:
+  def run(self):
+    write_test = _TextIOWritePerfTest(need_cleanup=False)
+    read_test = _TextIOReadPerfTest(input_folder=write_test.output_folder)
+    write_test.run()
+    read_test.run()
+
+
+class _TextIOWritePerfTest(LoadTest):
+  def __init__(self, need_cleanup=True):
+    super().__init__(WRITE_NAMESPACE)
+    self.need_cleanup = need_cleanup
+    self.test_options = self.pipeline.get_pipeline_options().view_as(
+        FileBasedIOTestOptions)
+    self.output_folder = FileSystems.join(
+        self.test_options.filename_prefix, str(uuid.uuid4()))
+
+  def test(self):
+    # first makedir if needed
+    _ = (
+        self.pipeline
+        | beam.Impulse()
+        | beam.ParDo(CreateFolderFn(self.output_folder)))
+
+    # write to text
+    _ = (
+        self.pipeline
+        | 'Produce rows' >> Read(
+            SyntheticSource(self.parse_synthetic_source_options()))
+        | 'Count records' >> beam.ParDo(CountMessages(self.metrics_namespace))
+        | 'Format' >> beam.ParDo(SyntheticRecordToStrFn())
```

Comment:
Why do we need to format these records?

Reply:
The bytes emitted from SyntheticSource are Key Value pairs with random bytes. We need to convert it to a text line (ideally a line of ascii chars and does not contain line break `\n`)